docs: Document On-Demand Runners and Tasks (#3297)
hashicorp/waypoint: website/content/docs/runner/on-demand-runner.mdx

---
layout: docs
page_title: On-Demand Runners
description: |-
  A brief overview about on-demand runners and their role in Waypoint.
---

# On-Demand Runners

On-demand runners allow a single Waypoint installation to scale very high by
leveraging the defined platform to spawn resources to execute Waypoint jobs
remotely. Through on-demand runner profiles and runner targeting, Waypoint can
even utilize resources from other platforms that Waypoint has access to. For
example, a local Kubernetes server install could be configured to use Amazon ECS
to execute jobs.

## Job Assignment

When a job is submitted to the server, if on-demand runners are enabled, it
will spawn an on-demand runner and the job is placed in a holding set. This is

> the same mechanism we use today for automatic CLI runners. When the on-demand

Reviewer: Given that these are stable docs rather than release notes, I’d consider dropping “today”.

```suggestion
the same mechanism we use for automatic CLI runners. When the on-demand
```

Author: Good call, agreed!

runner starts, it will execute the normal Waypoint runner code and connect back
to the server. On connection, the server will see the job in the holding set
and send it immediately to the new runner.

> The flow of information about the job will use the existing grpc runner APIs so
> all the input and output remain the same as today.

Reviewer: Similar to my comment above, “as today” is good framing for release notes but less so for “timeless” docs. Perhaps contrast it with the CLI-runner again?

```suggestion
The flow of information about the job uses the normal grpc runner APIs so all
the input and output remain the same as if the work were performed by the
CLI-runner.
```

runner profile.

## Official Waypoint Supported Task Plugin Platforms

When you use the `waypoint server install` command, we automatically set up a
default On-Demand runner profile for you to use with Waypoint Server. Like our
server install platform support, we also currently support these platforms
for launching on-demand runner tasks.

- [Kubernetes](/plugins/kubernetes#kubernetes-task)
- [HashiCorp/Nomad](/plugins/nomad#nomad-task)
- [AWS/ECS](/plugins/aws-ecs#aws-ecs-task)
- [Docker](/plugins/docker#docker-task)

## Task Plugin

Launching on-demand runners can be developed with the Task plugin. If there
is a platform that you wish to support while using Waypoint, we provide a
plugin system, similar to our build/deploy/release plugins, that lets you
define your own configuration for launching tasks on your platform of choice.

Plugins expect to fulfill the SDK interface with the following functions:

- `StartTask`
- `StopTask`

These functions are what will launch and stop the spawned resource with the
task launcher.

Task plugins also can accept `TaskLauncherConfig` configuration which users can
configure through an [on-demand runner profile](/docs/runner/profiles).

> ## Task Introspection

Reviewer: Similar to @krantzinator’s comment in Slack, perhaps we should ship this change alongside 0.9? Readers of the documentation might not know when to expect 0.9, and therefore when to expect this feature.

By default, the server will launch on-demand runners using the same OCI image
it was configured with. The on-demand configuration will also accept a specific
OCI image url to use instead. This allows operators the ability to prepackage
any Waypoint plugins into the image and use them in their configuration.

The official on-demand runner container image is built very similar to Waypoint
server, with some additional packages such as `Kaniko` for doing in-container
builds for jobs like the `Build operation`.

Users can configure Waypoint server to launch on-demand runners with a custom
built on-demand runner image. This image name must be set via the on-demand
runner profile.

## Official Waypoint Supported Task Plugin Platforms

When you use the `waypoint server install` command, we automatically set up a
default On-Demand runner profile for you to use with Waypoint Server. Like our
server install platform support, we also currently support these platforms
for launching on-demand runner tasks.

- [Kubernetes](/plugins/kubernetes#kubernetes-task)
- [HashiCorp/Nomad](/plugins/nomad#nomad-task)
- [AWS/ECS](/plugins/aws-ecs#aws-ecs-task)
- [Docker](/plugins/docker#docker-task)

> ## Task Plugin
> 
> Launching on-demand runners can be developed with the Task plugin. If there
> is a platform that you wish to support while using Waypoint, we provide a
> plugin system, similar to our build/deploy/release plugins, that lets you
> define your own configuration for launching tasks on your platform of choice.
> 
> Plugins expect to fulfill the SDK interface with the following functions:
> 
> - `StartTask`
> - `StopTask`
> 
> These functions are what will launch and stop the spawned resource with the
> task launcher.
> 
> Task plugins also can accept `TaskLauncherConfig` configuration which users can
> configure through an [on-demand runner profile](/docs/runner/profiles).

Reviewer: I wonder if we could reframe this section as “Adding on-demand runner support for other platforms”?

```suggestion
## Adding on-demand runner support for other platforms

On-demand runner support can be implemented for additional platforms
using the [plugin SDK](/docs/extending-waypoint).

Plugins expect to fulfill the SDK interface with the following functions:

- `StartTask`
- `StopTask`

These functions are what will launch and stop the spawned resource with the
task launcher.

Task plugins also can accept `TaskLauncherConfig` configuration which users can
configure through an [on-demand runner profile](/docs/runner/profiles).
```

Author: Good idea!
